### PR TITLE
dev/releases: PackageInfos to JSON script

### PR DESCRIPTION
**Pull request by ssiccha**
Please use the following template to submit a pull request, filling
in at least the "Text for release notes" and/or "Further details".
Thank You!

# Description

## Text for release notes 

If this pull request should be mentioned in the release notes, 
please provide a short description matching the style of the GAP
`CHANGES.md` file in the root directory.

## Further details

If necessary, please provide further details here.

# Checklist for pull request reviewers

- [ ] proper formatting

If your code contains kernel C code, run `clang-format` on it; the 
simplest way is to use `git clang-format`, e.g. like this (don't
forget to commit the resulting changes):

    git clang-format $(git merge-base HEAD master)

- [ ] usage of relevant labels

   1. either `release notes: not needed` or `release notes: to be added`
   2. at least one of the labels `bug` or `enhancement` or `new feature`
   3. for changes meant to be backported to `stable-4.X` add the `backport-to-4.X` label
   4. consider adding any of the labels `build system`, `documentation`, `kernel`, `library`, `tests`

- [ ] runnable tests
- [ ] lines changed in commits are sufficiently covered by the tests
- [ ] adequate pull request title
- [ ] well formulated text for release notes
- [ ] relevant documentation updates
- [ ] sensible comments in the code

### CODE REVIEW

File: dev/releases/make_archives.py [hunk 1]
```diff
@@ -205,11 +205,17 @@
     with open(filename+".sha256", 'w') as file:
         file.write(sha256file(filename))
 
+    # now create the file package-infos.json
+    subprocess.run(["../bin/gap.sh", "-r", "--quitonbreak",
```
Comment (fingolfin) [inline dev/releases/make_archives.py:209]: But at this point we already deleted all packages. Plus we never compiled the JSON package... so how can this work?

Comment (ssiccha) [reply dev/releases/make_archives.py]: Hm, I must have tested this with the GAP in my repositories root directory which has JSON compiled. :/

Comment (ssiccha) [reply dev/releases/make_archives.py]: I think I just pushed a fix. I'm testing it now.

Comment (ssiccha) [reply dev/releases/make_archives.py]: Ok, I have finally figured out where to compile json and call the new function. Let's see.

File: dev/releases/make_archives.py [hunk 2]
```diff
@@ -210,6 +227,7 @@
     with open(manifest_filename, 'w') as manifest:
         for archive in archives_to_create:
             manifest.write(f"{archive}\n")
+        manifest.write("package-infos.json\n")
```
Comment (fingolfin) [inline dev/releases/make_archives.py:230]: The filename here doesn't seem to match the one used above?

Comment (fingolfin) [reply dev/releases/make_archives.py:230]: No sorry, it does, I got confused

File: dev/releases/make_archives.py [hunk 3]
```diff
@@ -144,6 +145,22 @@
...
+    with working_directory(path_to_json_package):
+        subprocess.run(["./configure"], check=True)
+        subprocess.run(["make"], check=True)
+
+    subprocess.run(["./bin/gap.sh", "-r", "--quitonbreak",
+                   "../../dev/releases/PackageInfos-to-JSON.g"],
+                   check=True)
+
+    with working_directory(path_to_json_package):
+        subprocess.run(["rm", "-rf", "bin/", "Makefile"],
```
Comment (fingolfin) [inline dev/releases/make_archives.py:161]: Perhaps also call `make clean` before that, to also remove *.o files etc.?

File: dev/releases/make_archives.py (hunk 2 above)
Comment (ssiccha) [reply dev/releases/make_archives.py:230]: I think that's correct. The `PackageInfos-to-JSON.g` generates the file `package-infos.json`.

File: dev/releases/make_archives.py (hunk 3 above)
Comment (ssiccha) [reply dev/releases/make_archives.py:163]: Done.

**Comment by ssiccha** [comment]
Ha, now it worked.

**Comment by ssiccha** [comment]
@fingolfin Do I have to add a command to load all packages or does GAP parse all PackageInfos.g in the `pkg/` subfolder no matter whether they are loaded or not?

**Review by wilfwilson — COMMENTED** [review]
It's looking good!

File: dev/releases/make_archives.py [hunk 4]
```diff
@@ -144,6 +145,24 @@
...
+        subprocess.run(["./configure"], check=True)
+        subprocess.run(["make"], check=True)
+
+    subprocess.run(["./bin/gap.sh", "-r", "--quitonbreak",
+                   "../../dev/releases/PackageInfos-to-JSON.g"],
+                   check=True)
+
+    with working_directory(path_to_json_package):
+        subprocess.run(["make", "clean"],
+                       check=True)
```
Comment (fingolfin) [inline dev/releases/make_archives.py:162]:
```suggestion
        subprocess.run(["make", "clean"], check=True)
```

File: dev/releases/PackageInfos-to-JSON.g [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+InstallMethod(_GapToJsonStreamInternal,
+[IsOutputStream, IsObject],
+function(o, x)
+    PrintTo(o, "null");
+end);
+
+out := OutputTextFile("package-infos.json", false);
+GapToJsonStream(out, GAPInfo.PackagesInfo);;
+QUIT;
```
Comment (fingolfin) [inline dev/releases/PackageInfos-to-JSON.g:11]: How about printing to stdout instead? Then the code calling it has full control over where the output goes; and only one script needs to contain the name of the JSON file

File: dev/releases/make_archives.py [hunk 6]
```diff
@@ -144,6 +145,24 @@
...
+    # We first build the json package, then create the package-infos.json, then
+    # clean up the json package again.
+    path_to_json_package = glob.glob(f'{tmpdir}/{basename}/pkg/json*')[0]
+    with working_directory(path_to_json_package):
+        subprocess.run(["./configure"], check=True)
+        subprocess.run(["make"], check=True)
+
+    subprocess.run(["./bin/gap.sh", "-r", "--quitonbreak",
+                   "../../dev/releases/PackageInfos-to-JSON.g"],
+                   check=True)
```
Comment (fingolfin) [inline dev/releases/make_archives.py:158]: I am confused -- won't this put the JSON file into the root directory of the gap-4.11.1 directory? Shouldn't it be one level farther up for all of this to work? But then I think you said it all works now, so I must be missing something, but what?

File: dev/releases/make_archives.py [hunk 7]
```diff
@@ -144,6 +145,24 @@
...
+    # now create the file package-infos.json
+    # We first build the json package, then create the package-infos.json, then
+    # clean up the json package again.
+    path_to_json_package = glob.glob(f'{tmpdir}/{basename}/pkg/json*')[0]
+    with working_directory(path_to_json_package):
+        subprocess.run(["./configure"], check=True)
+        subprocess.run(["make"], check=True)
+
+    subprocess.run(["./bin/gap.sh", "-r", "--quitonbreak",
+                   "../../dev/releases/PackageInfos-to-JSON.g"],
```
Comment (fingolfin) [inline dev/releases/make_archives.py:157]: this hardcodes assumptions about the relative location of the current directory with respect to gap.git root directory. We can avoid this by moving the code which "removes unwanted files" to the end, then here you can just write:
```suggestion
                   "dev/releases/PackageInfos-to-JSON.g"],
```

File: dev/releases/make_archives.py (hunk 6 above)
Comment (ssiccha) [reply dev/releases/make_archives.py]: I fixed the script.

**Review by fingolfin — APPROVED** [review]

**Pull request merged**